pkg/resmgr/lib/memory/allocator.go

```diff
@@ -0,0 +1,771 @@
+
+	log.Debug("- finding initial zone")
+
+	if missing != 0 {
+		nodes, types := a.expand(req.zone, missing)
+		if types != missing && req.IsStrict() {
+			return fmt.Errorf("failed to find initial nodes of type %s", missing&^types)
+		}
+
+		req.zone |= nodes
+	}
+
+	if req.IsStrict() {
+		req.zone &= a.masks.nodes.byTypes[req.types]
+		missing = req.types &^ a.zoneType(req.zone)
+		if missing != 0 {
+			return fmt.Errorf("failed find initial nodes of type %s", missing)
+		}
+	} else {
+		if req.zone&a.masks.nodes.byTypes[req.types] != 0 {
+			req.zone &= a.masks.nodes.byTypes[req.types]
+		}
+	}
+
+	return nil
+}
+
+func (a *Allocator) ensureNormalMemory(req *Request) error {
+	//
+	// Make sure that request has some initial normal memory.
+	//
+	// We assume that we always have some normal DRAM present and therefore
+	// only force DRAM into the allowed types of non-strict requests.
+	//
+
+	if (req.zone & a.masks.nodes.normal) != 0 {
+		return nil
+	}
+
+	zone := req.zone
+	types := req.types & a.zoneType(a.masks.nodes.normal)
+	if types == 0 && !req.IsStrict() { // TODO(klihub): should we force this for strict, too ?
+		types |= TypeMaskDRAM
+	}
+
+	log.Debug("- ensuring normal memory for %s", zone)
+
+	for {
+		newNodes, newTypes := a.expand(zone, types)
+		if newNodes == 0 {
+			return fmt.Errorf("failed to find normal memory (of any type %s)", types)
+		}
+
+		zone |= newNodes
+		types |= newTypes
+
+		if (zone & a.masks.nodes.normal) != 0 {
+			req.zone = zone
+			req.types = types
+
+			return nil
+		}
+	}
+}
+
+func (a *Allocator) newOffer(req *Request, updates map[string]NodeMask) *Offer {
+	return &Offer{
+		a:       a,
+		req:     req,
+		updates: updates,
+		version: a.version,
+	}
+}
+
+func (a *Allocator) expand(zone NodeMask, types TypeMask) (NodeMask, TypeMask) {
+	var nodes NodeMask
+	if a.custom.ExpandZone != nil {
+		nodes = a.custom.ExpandZone(zone, types, &customAllocator{a}) &^ zone
+		types = a.zoneType(nodes)
+	} else {
+		nodes, types = a.defaultExpand(zone, types)
+	}
+
+	if nodes != 0 {
+		log.Debug("  + %s expanded by %s %s to %s", zoneName(nodes), types, nodes, zone|nodes)
+	}
+
+	return nodes, types
+}
+
+func (a *Allocator) defaultExpand(nodes NodeMask, types TypeMask) (NodeMask, TypeMask) {
+	var (
+		newNodes NodeMask
+		newTypes TypeMask
+	)
+
+	//
+	// This is our default zone/node expansion algorithm.
+	//
+	//   - for each type T
+	//     - for each node N in the zone
+	//       - for each distance for N
+	//         - take the closest set of new nodes of type T
+	//
+
+	types.Foreach(func(t Type) bool {
+		minDist := math.MaxInt
+		distIDs := map[int]NodeMask{}
+
+		a.ForeachNode(nodes, func(n *Node) bool {
+			n.ForeachDistance(func(dist int, distNodes NodeMask) bool {
+				if dist > minDist {
+					return false
+				}
+
+				extra := distNodes & a.masks.nodes.byTypes[t.Mask()] &^ nodes
+				if extra == 0 {
+					return dist < minDist // only continue if next set closer than minDist
```

Comment:
Not fully sure about this, so this is more of a question: is it possible that a later call into this for-callback would come with dist == minDist and with `extra != 0`? That is, it would still expand to some valid extra nodes if we would allow continuing the loop. If this is the case, maybe we should actually `return dist <= minDist`...

Reply:
Hmm... I'm not entirely sure why that would be better... but let me describe this in a bit more detail and then you can tell me whether it would make sense (and why) in the light of that if that is still the case.

The innermost `n.ForeachDistance()` iterates sets of nodes (NodeMasks) at the same distance by increasing distance from `n`. So if `dist == minDist`, it is guaranteed that we are right now considering the last possible set of nodes at minDist or closer distance, so there should be no reason to consider the next closest set of nodes, but... If `extra == 0` and `types` is a TypeMask of more than one type, then we might do other iterations (of the outermost loop) for another type `t`, and during that iteration we might pick up nodes of other type from the very same node set. In this light, is the original/suggestion still valid ?

I think regarding this implementation of expand one relevant question is: would it make more sense to ditch the current outermost loop, IOW type-by-type iteration, and simply iterate `nodes` by distance and neighbor set, since we `a.masks.nodes.byTypes` is set up so that you can look up nodes by any type mask, not just ones with a single bit set...

Reply:
@askervin I updated `defaultExpand()` along those lines described above.